iree/compiler/Conversion/LinalgToLLVM/FastExpConversion.cpp

```diff
@@ -0,0 +1,121 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/LLVMIR/LLVMTypes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace {
+
+// Fast polynomial approximation of exp(x) using its reduced range exp(y)
+// where y is in the range [0, ln(2)], let y = x - floor(x / ln(2)) * ln(2)
+// = x - k * ln(2), exp(x) = exp(y) * 2^k. exp(y) is computed with 4th degree
+// polyomial: exp(y) = c0 + c1 * y + c2 * y^2 + c3 * y^3 + c4 * y^4
+struct FastExpConversionPattern : public OpRewritePattern<LLVM::ExpOp> {
+  using OpRewritePattern<LLVM::ExpOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(LLVM::ExpOp op,
+                                PatternRewriter &rewriter) const override {
+    constexpr float ln2Const = 0.69314718055994529;
+    constexpr float ln2InvConst = 1.4426950408889634;
+    constexpr float cVaues[5] = {0.05924867, 0.15514645, 0.50308552, 0.99968939,
+                                 1.0000072153251447};
+    auto loc = op.getLoc();
+    Value x = op.getOperand();
+
+    auto floatType = LLVM::LLVMType::getFloatTy(rewriter.getContext());
+    auto i32Type = LLVM::LLVMType::getInt32Ty(rewriter.getContext());
+
+    Value ln2 = rewriter.create<LLVM::ConstantOp>(
+        loc, floatType, rewriter.getF32FloatAttr(ln2Const));
+    Value ln2Inv = rewriter.create<LLVM::ConstantOp>(
+        loc, floatType, rewriter.getF32FloatAttr(ln2InvConst));
+
+    // Compute reduced range input y = x - floor(x / ln(2)) * ln(2)
+    Value xL2Inv = rewriter.create<LLVM::FMulOp>(loc, floatType, x, ln2Inv);
+    Value kF32 = rewriter.create<LLVM::FFloorOp>(loc, floatType, xL2Inv);
+    Value kLn2 = rewriter.create<LLVM::FMulOp>(loc, floatType, kF32, ln2);
+    Value y = rewriter.create<LLVM::FSubOp>(loc, floatType, x, kLn2);
+
+    SmallVector<Value, 4> PConst(5);
+    for (int i = 0; i < 5; ++i) {
+      PConst[i] = rewriter.create<LLVM::ConstantOp>(
+          loc, floatType, rewriter.getF32FloatAttr(cVaues[i]));
+    }
+    // Evaluate exp(y) = sum(c[i] * y**i, i)
+    Value expY = rewriter.create<LLVM::FMulOp>(loc, floatType, y, PConst[0]);
+    expY = rewriter.create<LLVM::FAddOp>(loc, floatType, expY, PConst[1]);
+    expY = rewriter.create<LLVM::FMulOp>(loc, floatType, expY, y);
+    expY = rewriter.create<LLVM::FAddOp>(loc, floatType, expY, PConst[2]);
+    expY = rewriter.create<LLVM::FMulOp>(loc, floatType, expY, y);
+    expY = rewriter.create<LLVM::FAddOp>(loc, floatType, expY, PConst[3]);
+    expY = rewriter.create<LLVM::FMulOp>(loc, floatType, expY, y);
+    expY = rewriter.create<LLVM::FAddOp>(loc, floatType, expY, PConst[4]);
+
+    // Compute 2^k with integer bitshift, 2^k = (127 + k) << 23
```

Comment:
Sorry, I don't understand what is meant in this equation, `2^k = (127 + k) << 23`. The letter `k` appears on both sides of it, i just need help parsing what that means. A longer comment maybe?

Reply:
its exp2(k) computed with l-shift. aded a comment

Reply:
oh, thanks! i see what you're doing now - 23 is the number of mantissa bits, you're punning the float as an integer... so IIUC this relies on `k` being exactly an integral value. That's fairly reasonable as it was computed as the result of a `floor`, but I wonder what would happen for very large values which are not exactly representable as integers? i.e. absolute values larger than (1 / epsilon) ~= 1e+7 ? Are we safe from this case thanks to clamping of arguments somewhere? Or if you want to leave this as a TODO, that's fine too; there is a TODO below about float underflow/overflow, but here we are talking about much more 'tame' values (absolute values 1e+7 vs 1e+38). With clamping this is a non-issue because the exponential function is trivial so far away from 0, but without clamping it looks like there is a risk of this implementation of exponential producing surprising values for large absolute-value argument?  E.g. `exp(-1e+8f)` should return 0, but might return some weird value here ? maybe add a testcase in the unit test, possibly commented out for now if you want to handle it later?